src/librustc_codegen_ssa/back/write.rs

```diff
@@ -0,0 +1,1826 @@
+            emit_bc: false,
+            emit_bc_compressed: false,
+            emit_lto_bc: false,
+            emit_ir: false,
+            emit_asm: false,
+            emit_obj: false,
+            obj_is_bitcode: false,
+            embed_bitcode: false,
+            embed_bitcode_marker: false,
+            no_integrated_as: false,
+
+            verify_llvm_ir: false,
+            no_prepopulate_passes: false,
+            no_builtins: false,
+            time_passes: false,
+            vectorize_loop: false,
+            vectorize_slp: false,
+            merge_functions: false,
+            inline_threshold: None
+        }
+    }
+
+    fn set_flags(&mut self, sess: &Session, no_builtins: bool) {
+        self.verify_llvm_ir = sess.verify_llvm_ir();
+        self.no_prepopulate_passes = sess.opts.cg.no_prepopulate_passes;
+        self.no_builtins = no_builtins || sess.target.target.options.no_builtins;
+        self.time_passes = sess.time_passes();
+        self.inline_threshold = sess.opts.cg.inline_threshold;
+        self.obj_is_bitcode = sess.target.target.options.obj_is_bitcode ||
+                              sess.opts.debugging_opts.cross_lang_lto.enabled();
+        let embed_bitcode = sess.target.target.options.embed_bitcode ||
+                            sess.opts.debugging_opts.embed_bitcode;
+        if embed_bitcode {
+            match sess.opts.optimize {
+                config::OptLevel::No |
+                config::OptLevel::Less => {
+                    self.embed_bitcode_marker = embed_bitcode;
+                }
+                _ => self.embed_bitcode = embed_bitcode,
+            }
+        }
+
+        // Copy what clang does by turning on loop vectorization at O2 and
+        // slp vectorization at O3. Otherwise configure other optimization aspects
+        // of this pass manager builder.
+        // Turn off vectorization for emscripten, as it's not very well supported.
+        self.vectorize_loop = !sess.opts.cg.no_vectorize_loops &&
+                             (sess.opts.optimize == config::OptLevel::Default ||
+                              sess.opts.optimize == config::OptLevel::Aggressive) &&
+                             !sess.target.target.options.is_like_emscripten;
+
+        self.vectorize_slp = !sess.opts.cg.no_vectorize_slp &&
+                            sess.opts.optimize == config::OptLevel::Aggressive &&
+                            !sess.target.target.options.is_like_emscripten;
+
+        self.merge_functions = sess.opts.optimize == config::OptLevel::Default ||
+                               sess.opts.optimize == config::OptLevel::Aggressive;
+    }
+}
+
+/// Assembler name and command used by codegen when no_integrated_as is enabled
+pub struct AssemblerCommand {
+    name: PathBuf,
+    cmd: Command,
+}
+
+/// Additional resources used by optimize_and_codegen (not module specific)
+#[derive(Clone)]
+pub struct CodegenContext<B : WriteBackendMethods> {
```

Comment:
This reminds me the name `Codegen` in the context of the "back" module, is suboptimal.
Including in "optimize_and_codegen". I think we could call this step "emitting machine code/a binary", so maybe `optimize_and_emit`? `OptimizeAndEmitContext` is not great, but more descriptive.

cc @michaelwoerister @rkruppe @alexcrichton 

Also, ` : ` should be `: `.